### PR TITLE
Fix #437: Compute initial Y scaling based on data in view

### LINKED ISSUE #437: Improve initial Y scaling
* Initial scaling should be computed based on the currently visible measurements
* To have a better view on the scale or the diffs, it would be better to start with Y=0

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@liufengyun I could not manage to test it locally. I tried to follow the `web/README` but it failed. Could you try it on your end?

File: web/main.js [hunk 1]
```diff
@@ -175,8 +175,17 @@ window.showCommit = function () {
...
+
     options.xaxis.autorange = false;
-    options.xaxis.range = [data[0].x.length - 100, data[0].x.length];
+    options.xaxis.range = [x_data.length - 100, x_data.length];
+
+    var max_y = 0;
+    for (let i = y_data.length - 100; i < y_data.length; i++) {
+      max_y = Math.max(max_y, y_data[i]);
+    }
+    options.yaxis.range = [0, max_y];
```
Comment (liufengyun) [inline web/main.js:188]: Cool 👍 

Do we need to set `options.yaxis.autorange`:

```
options.yaxis.range = false
```

**Comment by liufengyun** [comment]
> @liufengyun I could not manage to test it locally. I tried to follow the `web/README` but it failed. Could you try it on your end?

I'll run it locally. BTW, what kind of failure do you get?

**Comment by liufengyun** [comment]
@nicolasstucki I've pushed your changes here: https://liufengyun.github.io/bench/

(may need clear cache)

File: web/main.js (hunk 1 above)
Comment (liufengyun) [reply web/main.js:188]: Just tried, it's necessary.

**Comment by nicolasstucki** [comment]
Cool. I gave it an extra margin on top.

**Comment by liufengyun** [comment]
It seems starting from 0 is not good --- maybe use `min_y * 0.8`?

**Comment by nicolasstucki** [comment]
It faild with 

```scala
(base) ➜  web git:(fix-#437) ✗ npm run watch

> bench-web@ watch /Users/nicolasstucki/GitHub/bench/web
> browser-sync start --server --files './main.js'

sh: browser-sync: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! bench-web@ watch: `browser-sync start --server --files './main.js'`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the bench-web@ watch script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/nicolasstucki/.npm/_logs/2020-08-27T08_12_09_207Z-debug.log
```


```scala
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/local/Cellar/node/12.6.0/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'run',
1 verbose cli   'watch'
1 verbose cli ]
2 info using npm@6.14.5
3 info using node@v12.6.0
4 verbose run-script [ 'prewatch', 'watch', 'postwatch' ]
5 info lifecycle bench-web@~prewatch: bench-web@
6 info lifecycle bench-web@~watch: bench-web@
7 verbose lifecycle bench-web@~watch: unsafe-perm in lifecycle true
8 verbose lifecycle bench-web@~watch: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/nicolasstucki/GitHub/bench/web/node_modules/.bin:/Users/nicolasstucki/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/nicolasstucki/anaconda3/bin:/Library/TeX/texbin:/usr/local/share/dotnet:/opt/X11/bin:~/.dotnet/tools:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/nicolasstucki/anaconda3/condabin:/Users/nicolasstucki/.cargo/bin:/Applications/Visual Studio Code.app/Contents/Resources/app/bin:/Applications/scala-2.12.6/bin:/Applications/Visual Studio Code.app/Contents/Resources/app/bin:/Applications/scala-2.12.6/bin
9 verbose lifecycle bench-web@~watch: CWD: /Users/nicolasstucki/GitHub/bench/web
10 silly lifecycle bench-web@~watch: Args: [ '-c', "browser-sync start --server --files './main.js'" ]
11 info lifecycle bench-web@~watch: Failed to exec watch script
12 verbose stack Error: bench-web@ watch: `browser-sync start --server --files './main.js'`
12 verbose stack spawn ENOENT
12 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:48:18)
12 verbose stack     at ChildProcess.emit (events.js:203:13)
12 verbose stack     at maybeClose (internal/child_process.js:1021:16)
12 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
13 verbose pkgid bench-web@
14 verbose cwd /Users/nicolasstucki/GitHub/bench/web
15 verbose Darwin 19.6.0
16 verbose argv "/usr/local/Cellar/node/12.6.0/bin/node" "/usr/local/bin/npm" "run" "watch"
17 verbose node v12.6.0
18 verbose npm  v6.14.5
19 error code ELIFECYCLE
20 error syscall spawn
21 error file sh
22 error errno ENOENT
23 error bench-web@ watch: `browser-sync start --server --files './main.js'`
23 error spawn ENOENT
24 error Failed at the bench-web@ watch script.
24 error This is probably not a problem with npm. There is likely additional logging output above.
25 verbose exit [ 1, true ]
```

**Comment by liufengyun** [comment]
@nicolasstucki Have you run `npm install` to install `browser-sync` first?

**Comment by nicolasstucki** [comment]
Now it worked

**Review by liufengyun — APPROVED** [review]
Nice work 👍 

It seems starting from 0 is not good enough --- maybe use `min_y * 0.8`?

**Comment by nicolasstucki** [comment]
Y =0 gives a better intuition of the scale of the difference. Also, it is easy to zoom in but harder to zoom out to include the 0.

**Comment by liufengyun** [comment]
I just pushed the latest to https://liufengyun.github.io/bench/

It seems `1.1` is too small, maybe `1.3`?

**Comment by liufengyun** [comment]
> Y =0 gives a better intuition of the scale of the difference. Also, it is easy to zoom in but harder to zoom out to include the 0.

OK, agreed.

**Comment by liufengyun** [comment]
Thanks, now it is deployed https://github.com/lampepfl/bench/commit/897d3b0612d40bbd1571216e488653f0c23f885b

**Comment by smarter** [comment]
>Y =0 gives a better intuition of the scale of the difference

I'm not so sure, when I look at the main dotty graph on http://dotty-bench.epfl.ch/, it looks almost flat, it's much easier to see small differences when I make the Y axis range from 40k to 50k instead of 0 to 50k, and this doesn't hide any data so zooming out isn't needed much (also there's a button that can be clicked to automatically zoom out to the full graph)

**Comment by smarter** [comment]
In fact, right now almost all graphs look flat, if I quickly scroll through the page it's hard to see any trend, if we zoomed the Y axis then we see things much beter imho.

**Comment by nicolasstucki** [comment]
Plotly provides a simple tool to zoom in but not to zoom out. If we zoom in by default we also lose the scale of any improvement and highlight the noise in the benchmarks.

**Comment by liufengyun** [comment]
One advantage of starting from 0 is to quickly identify noticeable regression/speedup, instead of noises.

From the last meeting on performance infrastructure, it seems 1%-3% noise is unavoidable.

**Comment by liufengyun** [comment]
I didn't something like set starting Y from `Math.max(min - max + min, 0)`:

https://github.com/liufengyun/bench/commit/893af395f7731517eb7d3887973b0acba7f4d15a

The demo is at https://liufengyun.github.io/bench/ (may need to clear cache)

**Pull request merged**